migrator/data/DataImporter.java

```diff
@@ -239,50 +241,64 @@ void run() {
         }
 
         private void commitBatch() {
+            assert originalToBufferedIDs.keySet().containsAll(incompleteIDs);
             transaction.commit();
-            transaction.committedIIDs().forEachRemaining(pair ->
-                    idMapper.put(bufferedToOriginalIds.get(pair.first()), pair.second())
-            );
-            bufferedToOriginalIds.clear();
-            originalToBufferedIds.clear();
+            transaction.committedIIDs().forEachRemaining(pair -> {
+                String originalID = bufferedToOriginalIDs.get(pair.first());
+                conceptTracker.recordMapping(originalID, pair.second());
+                if (incompleteIDs.contains(originalID)) conceptTracker.recordIncomplete(originalID);
+            });
+            bufferedToOriginalIDs.clear();
+            originalToBufferedIDs.clear();
+            completedIDs.forEach(conceptTracker::deleteIncomplete);
+        }
```

Comment:
a commit  batch has to track completed and incomplete IDs, as well as ID mappings